File: kotlin-node/karakum/src/jsMain/resources/plugins/convertOverriddenPropertySignature.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

const overriddenProps = {
    // child_process
    "ChildProcessByStdio": {
        "stdin": "boundStdin",
        "stdout": "boundStdout",
        "stderr": "boundStderr",
        "stdio": "boundStdio",
    },
    "ChildProcessWithoutNullStreams": {
        "stdin": "requiredStdin",
        "stdout": "requiredStdout",
        "stderr": "requiredStderr",
        "stdio": "requiredStdio",
    },
    "ExecFileOptions": {
        "signal": null,
    },
    "ExecFileSyncOptionsWithBufferEncoding": {
        "encoding": "bufferEncoding",
    },
    "ExecFileSyncOptionsWithStringEncoding": {
        "encoding": "stringEncoding",
    },
    "ExecSyncOptionsWithBufferEncoding": {
        "encoding": "bufferEncoding",
    },
    "ExecSyncOptionsWithStringEncoding": {
        "encoding": "stringEncoding",
    },
    "SpawnSyncOptionsWithBufferEncoding": {
        "encoding": "bufferEncoding",
    },
    "SpawnSyncOptionsWithStringEncoding": {
        "encoding": "stringEncoding",
    },
    "SpawnOptionsWithStdioTuple": {
        "stdio": "stdioTuple",
    },
    "SpawnOptionsWithoutStdio": {
        "stdio": null,
    },

    // crypto
    "GeneratePrimeOptionsArrayBuffer": {
        "bigint": null,
    },
    "GeneratePrimeOptionsBigInt": {
        "bigint": "requiredBigInt",
    },

    // dns
    "LookupAllOptions": {
        "all": "requiredAll",
    },
    "LookupOneOptions": {
        "all": null,
    },
    "ResolveWithTtlOptions": {
        "ttl": null,
    },

    // fs
    "GlobOptions": {
        "exclude": "globExclude",
    },
    "GlobOptionsWithFileTypes": {
        "exclude": "excludeWithFileTypes",
        "withFileTypes": "requiredWithFileTypes"
    },
    "GlobOptionsWithoutFileTypes": {
        "exclude": "excludeWithoutFileTypes",
        "withFileTypes": null
    },

    // http
    "AgentOptions": {
        "keepAlive": null
    },

    // http2
    "Http2Stream": {
        "closed": null,
        "destroyed": "readonlyDestroyed"
    },

    // tls
    "TlsOptions": {
        "sessionTimeout": null,
        "ticketKeys": null,
    },

    // url
    "UrlWithParsedQuery": {
        "query": "parsedQuery",
    },
    "UrlWithStringQuery": {
        "query": "stringQuery",
    },
}

export default function (node, context, render) {
    if (
        ts.isPropertySignature(node)
        && ts.isInterfaceDeclaration(node.parent)
        && node.parent.name.text in overriddenProps
        && ts.isIdentifier(node.name)
        && node.name.text in overriddenProps[node.parent.name.text]
    ) {
        const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)

        const modifier = readonly
            ? "val "
            : "var "

        const name = render(node.name)

        const isOptional = Boolean(node.questionToken)

        const type = karakum.renderNullable(node.type, isOptional, context, render)

        const overrideConfig = overriddenProps[node.parent.name.text][node.name.text]

        const hasAlias = overrideConfig !== null

        if (hasAlias) {
            let alias
            let parentType

            if (Array.isArray(overrideConfig)) {
                [alias, parentType] = overrideConfig
            } else {
                alias = overrideConfig
            }

            const jsName = `@JsName("${name}")`
            const aliasSignature = `${modifier}${alias}: ${type}`

            if (parentType) {
                const deprecation = `@Deprecated(message = "use ${alias}", level = DeprecationLevel.HIDDEN)`
                const signature = `override ${modifier}${name}: ${parentType}`

                return [
                    deprecation,
                    signature,
                    jsName,
                    aliasSignature,
                ].join("\n")
            } else {
                return [
                    jsName,
                    aliasSignature,
                ].join("\n")
            }
        }

        return `override ${modifier}${name}: ${type}`
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertOverriddenPropertySignature.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.getParentOrNull
import node.karakum.plugins.SignatureOverrideConfig.Alias
import node.karakum.plugins.SignatureOverrideConfig.Preserve
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isPropertySignature

private sealed interface SignatureOverrideConfig {
    object Preserve : SignatureOverrideConfig
    class Alias(val alias: String) : SignatureOverrideConfig
}

private val overriddenProps = mapOf(
    // child_process
    "ChildProcessByStdio" to mapOf(
        "stdin" to Alias("boundStdin"),
        "stdout" to Alias("boundStdout"),
        "stderr" to Alias("boundStderr"),
        "stdio" to Alias("boundStdio"),
    ),
    "ChildProcessWithoutNullStreams" to mapOf(
        "stdin" to Alias("requiredStdin"),
        "stdout" to Alias("requiredStdout"),
        "stderr" to Alias("requiredStderr"),
        "stdio" to Alias("requiredStdio"),
    ),
    "ExecFileOptions" to mapOf(
        "signal" to Preserve,
    ),
    "ExecFileSyncOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("bufferEncoding"),
    ),
    "ExecFileSyncOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "ExecSyncOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("bufferEncoding"),
    ),
    "ExecSyncOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "ExecFileOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("bufferEncoding"),
    ),
    "ExecFileOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "ExecOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("bufferEncoding"),
    ),
    "ExecOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "SpawnSyncOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("bufferEncoding"),
    ),
    "SpawnSyncOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "SpawnOptionsWithStdioTuple" to mapOf(
        "stdio" to Alias("stdioTuple"),
    ),
    "SpawnOptionsWithoutStdio" to mapOf(
        "stdio" to Preserve,
    ),

    // crypto
    "GeneratePrimeOptionsArrayBuffer" to mapOf(
        "bigint" to Preserve,
    ),
    "GeneratePrimeOptionsBigInt" to mapOf(
        "bigint" to Alias("requiredBigInt"),
    ),
    "OneShotDigestOptionsWithBufferEncoding" to mapOf(
        "outputEncoding" to Alias("bufferOutputEncoding"),
    ),
    "OneShotDigestOptionsWithStringEncoding" to mapOf(
        "outputEncoding" to Alias("stringOutputEncoding"),
    ),

    // dns
    "LookupAllOptions" to mapOf(
        "all" to Alias("requiredAll"),
    ),
    "LookupOneOptions" to mapOf(
        "all" to Preserve,
    ),
    "ResolveWithTtlOptions" to mapOf(
        "ttl" to Preserve,
    ),

    // fs
    "GlobOptions" to mapOf(
        "exclude" to Alias("globExclude"),
    ),
    "GlobOptionsWithFileTypes" to mapOf(
        "exclude" to Alias("excludeWithFileTypes"),
        "withFileTypes" to Alias("requiredWithFileTypes")
    ),
    "GlobOptionsWithoutFileTypes" to mapOf(
        "exclude" to Alias("excludeWithoutFileTypes"),
        "withFileTypes" to Preserve
    ),
    "WatchOptionsWithBufferEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),
    "WatchOptionsWithStringEncoding" to mapOf(
        "encoding" to Alias("stringEncoding"),
    ),

    // http
    "AgentOptions" to mapOf(
        "keepAlive" to Preserve
    ),

    // http2
    "Http2Stream" to mapOf(
        "closed" to Preserve,
        "destroyed" to Alias("readonlyDestroyed")
    ),

    // tls
    "TlsOptions" to mapOf(
        "sessionTimeout" to Preserve,
        "ticketKeys" to Preserve,
    ),

    // url
    "UrlWithParsedQuery" to mapOf(
        "query" to Alias("parsedQuery"),
    ),
    "UrlWithStringQuery" to mapOf(
        "query" to Alias("stringQuery"),
    ),
)

val convertOverriddenPropertySignature = createPlugin { node, context, render ->
    nullable {
        ensure(isPropertySignature(node))

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))

        val interfaceOverrides = ensureNotNull(overriddenProps[interfaceNode.name.text])

        val propertyName = node.name
        ensure(isIdentifier(propertyName))
        val propertyOverride = ensureNotNull(interfaceOverrides[propertyName.text])

        val readonly = node.modifiers?.asArray()?.find { it.kind == SyntaxKind.ReadonlyKeyword }

        val modifier = if (readonly != null) "val " else "var "

        val name = render(node.name)

        val isOptional = node.questionToken != null

        val type = renderNullable(node.type, isOptional, context, render)

        when (propertyOverride) {
            is Alias -> {
                val alias = propertyOverride.alias
                val jsName = "@JsName(\"$name\")"
                val aliasSignature = "${modifier}${alias}: $type"

                arrayOf(
                    jsName,
                    aliasSignature,
                ).joinToString("\n")
            }
            Preserve -> "override ${modifier}${name}: $type"
        }
    }
}
